### PR TITLE
Enhance export system

### PR DESCRIPTION
**Pull request by ospencer**
This PR introduces required exports for values and types in modules.

New syntax includes:
```
export let x = ...
export foreign wasm x ...
export data Foo = ...

export x
export x as y
export x, y as z

export Foo, Bar
export Foo, Bar as Baz

export *
export * except x, y, Foo
```

### CODE REVIEW

File: src/typed/typemod.ml [hunk 1]
```diff
@@ -284,60 +284,126 @@ let enrich_type_decls anchor decls oldenv newenv =
...
+        new_env, List.rev sigs @ signatures, statement::statements
+      | PTopExportData ex -> 
+        let sigs = process_export_data env ex loc in 
+        env, List.rev sigs @ signatures, statements
+      | PTopForeign(e, d) -> 
+        let new_env, signature, statement = process_foreign env e d loc in
+        let signatures = match signature with Some(s) -> s::signatures | None -> signatures in
+        new_env, signatures, statement::statements
+      | PTopData(e, d) -> 
+        let new_env, sigs, statement = process_datas env e [d] loc in
```
Comment (ospencer) [inline src/typed/typemod.ml]: One thing to point out here: the compiler was assuming that all data declarations were mutually recursive. I set this up to make it easy to parse mutually recursive datas and put them together into one `PTopData`. There will be a follow-up PR that implements that.

File: src/grain-stdlib/lists.grlib [hunk 2]
```diff
@@ -62,3 +62,5 @@ let rec map = (f, l) => {
     | Cons(hd, tl) => Cons(f(hd), map(f, tl))
   }
 };
+
+export *;
```
Comment (belph) [inline src/grain-stdlib/lists.grlib]: I feel like the style should be:
```
<imports/exports>

<code>
```

File: src/middle_end/linearize.ml [hunk 3]
```diff
@@ -176,12 +176,12 @@ let rec transl_imm (({exp_desc; exp_loc=loc; exp_env=env; _} as e) : expression)
...
 
-and bind_patts ?toplevel:(toplevel=false) (exp_id : Ident.t) (patts : pattern list) : anf_bind list =
+and bind_patts ?exported:(exported=false) (exp_id : Ident.t) (patts : pattern list) : anf_bind list =
   let postprocess_item cur acc =
     match cur with
     | None -> acc
     | Some(ident, (src, idx), extras) ->
-      let bind = if toplevel then
+      let bind = if exported then
           BLetGlobal(Nonrecursive, [ident, Comp.tuple_get (Int32.of_int idx) (Imm.id src)])
```
Comment (belph) [inline src/middle_end/linearize.ml]: The semantics of this have shifted from its original design...while this in practice does exactly what you want, I would love to see this renamed from `BLetGlobal` to `BLetExport`.

File: src/parsing/ast_helper.ml [hunk 4]
```diff
@@ -115,10 +115,20 @@ module Top = struct
...
   let import ?loc i = mk ?loc (PTopImport i)
-  let data ?loc d = mk ?loc (PTopData d)
-  let let_ ?loc r vb = mk ?loc (PTopLet(r, vb))
+  let foreign ?loc e d = mk ?loc (PTopForeign (e, d))
+  let data ?loc e d = mk ?loc (PTopData (e, d))
+  let let_ ?loc e r vb = mk ?loc (PTopLet(e, r, vb))
+  let export ?loc e = mk ?loc (PTopExport e)
+
+  let mk_export_list e =
+    List.map (fun name ->
```
Comment (belph) [inline src/parsing/ast_helper.ml]: This should be moved into the parser, as we discussed offline

File: src/parsing/ast_iterator.ml [hunk 5]
```diff
@@ -106,6 +108,26 @@ module I = struct
     iter_loc sub imod
 end
 
+module EX = struct
```
Comment (belph) [inline src/parsing/ast_iterator.ml]: `EX` -> `Ex`

File: src/parsing/ast_iterator.ml [hunk 6]
```diff
@@ -106,6 +108,26 @@ module I = struct
...
+    ) exports
+  let iter_export_all sub excepts =
+    List.iter (fun except ->
+      match except with
+      | ExportExceptData(name)
+      | ExportExceptValue(name) -> iter_loc sub name
+    ) excepts
+end
+
+module EXD = struct
```
Comment (belph) [inline src/parsing/ast_iterator.ml]: `EXD` -> `ExD` (or just `ED`)

File: src/parsing/ast_mapper.ml [hunk 7]
```diff
@@ -117,6 +119,28 @@ module I = struct
     mk ~loc (map_loc sub imod)
 end
 
+module EX = struct
```
Comment (belph) [inline src/parsing/ast_mapper.ml]: `EX` -> `Ex`

File: src/parsing/lexer.mll [hunk 8]
```diff
@@ -125,8 +128,8 @@ rule token = parse
   | "<" { LCARET }
   | ">" { RCARET }
   | "+" { PLUS }
-  | "-" { MINUS }
-  | "*" { TIMES }
+  | "-" { DASH }
```
Comment (belph) [inline src/parsing/lexer.mll]: :thinking:

File: src/typed/typedtreeIter.ml [hunk 9]
```diff
@@ -163,7 +163,7 @@ end = struct
       | TExpIdent _
       | TExpConstant _ -> ()
       | TExpLet(recflag, binds, body) ->
-        iter_bindings recflag binds;
+        iter_bindings Nonexported recflag binds;
```
Comment (belph) [inline src/typed/typedtreeIter.ml]: Doesn't sit well with me (stylistically), but I get it.

**Pull request merged**